### PR TITLE
Enforce config values having correct types before passing to gen_smtp

### LINKED ISSUE #161: Can't customize SMTP port
I pretty much have the same use case and issue as #129.

My config and the error I'm seeing are below.

My config:

```elixir
config :identity_context, IdentityContext.Mailer,
  adapter: Swoosh.Adapters.SMTP,
  password: { :system, "SMTP_PASSWORD" },
  port: { :system, "SMTP_PORT" },
  relay: { :system, "SMTP_HOST" },
  username: { :system, "SMTP_USERNAME" }
```

My error:

```
{:error,
 {:retries_exceeded,
  {:network_failure, 'app.mailhog', {:error, :econnrefused}}}}
```

I had a feeling it might be because the port is coming in from the config as a string and `gen_smtp` might not handle this properly. When I switched to a hardcoded integer in my config I got the following error:

```
** (FunctionClauseError) no function clause matching in :gen_smtp_client.to_string/1    
    
    The following arguments were given to :gen_smtp_client.to_string/1:
    
        # 1
        nil
    
    (gen_smtp) /usr/src/app/deps/gen_smtp/src/gen_smtp_client.erl:338: :gen_smtp_client.to_string/1
    (gen_smtp) /usr/src/app/deps/gen_smtp/src/gen_smtp_client.erl:320: :gen_smtp_client.try_AUTH/3
    (gen_smtp) /usr/src/app/deps/gen_smtp/src/gen_smtp_client.erl:204: :gen_smtp_client.do_smtp_session/3
    (gen_smtp) /usr/src/app/deps/gen_smtp/src/gen_smtp_client.erl:143: :gen_smtp_client.try_smtp_sessions/4
    (swoosh) lib/swoosh/adapters/smtp.ex:43: Swoosh.Adapters.SMTP.deliver/2
    (identity_context) lib/identity_context/registration/command.ex:31: IdentityContext.Registration.Command.send_welcome_email/1
```

### PR DESCRIPTION
**Pull request by princemaple**
Fixes #161

### CODE REVIEW

File: lib/swoosh/adapters/smtp.ex [hunk 1]
```diff
@@ -40,13 +41,51 @@ if Code.ensure_loaded?(:gen_smtp_client) do
...
         receipt when is_binary(receipt) -> {:ok, receipt}
         {:error, type, message} -> {:error, {type, message}}
         {:error, reason} -> {:error, reason}
       end
     end
 
+    @config_transformations %{
+      port: &String.to_integer/1,
+      retries: &String.to_integer/1,
+      ssl: {&String.to_existing_atom/1, [true, false]},
```
Comment (stevedomin) [inline lib/swoosh/adapters/smtp.ex]: why are you using `to_existing_atom/1` here? (just curious)

**Comment by stevedomin** [comment]
@princemaple looks great, just a minor suggestion if you find the time.

**Comment by princemaple** [comment]
Yep, was just trying to see if it works first.

File: lib/swoosh/adapters/smtp.ex (hunk 1 above)
Comment (princemaple) [reply lib/swoosh/adapters/smtp.ex]: So the user cannot misspell the config and create a new one.
While it doesn't really matter, as it's caught immediately, I always use `to_existing_atom` if `to_atom` is not absolutely needed.

Comment (stevedomin) [reply lib/swoosh/adapters/smtp.ex]: 👍

**Comment by princemaple** [comment]
OK this seems to work well. I tested it by setting up a new test project, traced through with IEx.break! and pry. I'll add some tests and merge.

File: lib/swoosh/adapters/smtp.ex (hunk 1 above)
Comment (princemaple) [reply lib/swoosh/adapters/smtp.ex]: I changed this to `to_atom` in the end:

- this is not from user input, should be safe to use `to_atom`
- using `to_existing_atom` raises earlier than my valid option check, troublesome

**Comment by princemaple** [comment]
@stevedomin please take another look.

**Comment by stevedomin** [comment]
Thanks @princemaple, this is perfect!

**Pull request merged**